src/ui_parts/viewport.gd

```diff
@@ -71,32 +41,43 @@ func center_frame() -> void:
 func _unhandled_input(event: InputEvent) -> void:
 	if not event is InputEventMouseMotion or event.button_mask != 0:
 		view.queue_redraw()
+	
 	if event is InputEventMouseMotion and\
 	event.button_mask in [MOUSE_BUTTON_MASK_LEFT, MOUSE_BUTTON_MASK_MIDDLE]:
 		set_view(view.position - event.relative)
 	
 	if event is InputEventPanGesture:
 		if event.ctrl_pressed:
-			zoom_level *= 1 + event.delta.y / 2
+			%ZoomMenu.zoom_level *= 1 + event.delta.y / 2
 		else:
 			set_view(view.position + event.delta * 32)
 	
 	if event is InputEventMagnifyGesture:
-		zoom_level *= event.factor
+		%ZoomMenu.zoom_level *= event.factor
 	
 	if event is InputEventMouseButton and event.is_pressed():
 		match event.button_index:
+			MOUSE_BUTTON_WHEEL_UP when GlobalSettings.invert_zoom:
+				%ZoomMenu.zoom_out()
 			MOUSE_BUTTON_WHEEL_UP:
-				if GlobalSettings.invert_zoom:
-					zoom_out()
-				else:
-					zoom_in()
+				%ZoomMenu.zoom_in()
+			MOUSE_BUTTON_WHEEL_DOWN when GlobalSettings.invert_zoom:
+				%ZoomMenu.zoom_in()
 			MOUSE_BUTTON_WHEEL_DOWN:
-				if GlobalSettings.invert_zoom:
-					zoom_in()
-				else:
-					zoom_out()
+				%ZoomMenu.zoom_out()
 	
 	if event is InputEventMouseButton:
 		if event.ctrl_pressed:
 			pass
+
+
+func _on_zoom_changed(zoom_level: float) -> void:
+	var old_size_2d_override := size_2d_override
+	size_2d_override = size / zoom_level
+	set_view(view.position + (old_size_2d_override - size_2d_override) / 2.0)
+	display.material.set_shader_parameter(&"uv_scale",
+			nearest_po2(int(zoom_level * 32)) / 32.0)
+	update_view_limits()
+	controls.zoom = zoom_level
+	display_texture.zoom = zoom_level
```

Comment:
Can you check if changing the zoom variable in these two nodes can be centralized to the new zoom widget instead?

Reply:
Like this?  

```gdscript
# handles_manager.gd

# ...

@onready var zoom_menu: HBoxContainer = %ZoomMenu

# ...

func _draw() -> void:
	# error because doesn't know the zoom_level type
	var thickness := 0.85 / zoom_menu.zoom_level
```

I thought about two solutions for this:  
1- We could give a type to the `ZoomMenu` instead of using `HBoxContainer`.  
2- We could have a "temporary/soft settings" to hold the zoom information instead of letting the widget/display have it.  

```gdscript
# temp_settings.gd
class_name TempSettings

signal zoom_changed(zoom: float)

var zoom: float = 1.0:
  set(value):
    zoom = clampf(value, MIN_ZOOM, MAX_ZOOM)
    zoom_changed(zoom)
```

(More tests would be needed in this case)